### PR TITLE
Sidebar active tabs

### LINKED ISSUE #146: Websites should remain active even when a tab is closed
Original report:

<div type='discussions-op-text'>

<sup>Originally posted by **juush** November 14, 2022</sup>
Would be nice if you could "minimize" the side bar tabs.
I'd like to have pandora playing from the browser but not have to have the tab open the entire time!</div>

**Tl;Dr:** Websites in the sidebar should remain active even when they are closed / hidden. This might need some significant engineering to get it to work correctly, so I will probably be the one to take it on.

### PR DESCRIPTION
**Pull request by PressJump**
Fixes #146 Implementing Sidebar Background Active Tabs

### CODE REVIEW

**Comment by PressJump** [comment]
## QA Requirements

- [x] Sidebar

  - [x] The sidebar exists
  - [x] The settings and addons button are visible down the bottom
  - [x] The bookmarks, history and synced tabs icons are in the sidebar
  - [x] The sidebar can be resized properly
  - [x] An [installed extension](https://addons.mozilla.org/en-US/firefox/addon/sidebar-note/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search) will appear in the sidebar
  - [x] Sidebar-Tabs is Loaded and has a button visible down the bottom
  - [x] Sidebar-Tabs can add a sidebar item to the sidebar
  - [x] Sidebar-Tabs can remove a sidebar item to the sidebar
  - [x] The sidebar keeps tabs active while closed when the preference `pulse.sidebar.keeptabsactive.enabled` is set to true
  - [x] The sidebar overwrites tabs when the preference `pulse.sidebar.keeptabsactive.enabled` is set to false



- [x] Extensions

  - [x] Tabliss is loading correctly
  - [x] Tabliss does not have the webextention symbol in the title bar
  - [x] uBlock origin is loading correctly
  - [x] QRCode generator is loading correctly

- [x] About dialog

  - [x] The new about dialog is present
  - [x] The dialog has not significantly changed

- [x] Pocket

  - [x] Pocket is not present in the browser

- [x] Preferences

  - [x] All tabs in preferences open
  - [x] The "More from {Vendor}" option is not present
  - [x] The sidebar preferences are present and work as expected

- [x] Welcome experience

  - [x] The import button opens the import wizard
  - [x] The theme picker is present and works as expected
  - [x] The search engine picker is present and works as expected

- [x] Pulse theme

  - [x] The tabs appear to be consistent

- [x] Mozilla theme

  - [x] The sidebar tabs are correct

- [x] Useragent
  - [x] The user agent is reporting as Mozilla Firefox

**Comment by PressJump** [comment]
Currently browsers do not keep up with new sidebar sizing on 107, and removed shortcuts still exist in the background. With changes coming with Sidebar Active Tabs, the QA checks will need an update too. In draft for now due to this.

**Comment by PressJump** [comment]
Removed shortcuts now get removed after restart of the browser. This is the case with every other loaded tab as well. Persistence is only kept when keepTabsActive is set to false.

**Comment by PressJump** [comment]
Welcome options for sidebar will be available on next PR.

File: .github/PULL_REQUEST_TEMPLATE/ff_update.md [hunk 1]
```diff
@@ -14,7 +14,8 @@ This bumps Firefox up to `{NEW_VERSION}` for security and performance reasons. R
   - [ ] Sidebar-Tabs is Loaded and has a button visible down the bottom
   - [ ] Sidebar-Tabs can add a sidebar item to the sidebar
   - [ ] Sidebar-Tabs can remove a sidebar item to the sidebar
-
+  - [ ] The sidebar keeps tabs active while closed when the preference `pulse.sidebar.keeptabsactive.enabled` is set to true
```
Comment (trickypr) [inline .github/PULL_REQUEST_TEMPLATE/ff_update.md]: We do not care about preference keys, we only care about user facing options.

File: src/browser/base/content/browser-sidebar-js.patch [hunk 2]
```diff
@@ -134,7 +144,37 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
      this._inited = true;
  
      Services.obs.addObserver(this, "intl:app-locales-changed");
-@@ -159,17 +229,26 @@ var SidebarUI = {
+ 
+     this._initDeferred.resolve();
++    
++    //check if element by id sidebar exists to bool var
++
++    var sidebar = document.getElementById("sidebar");
```
Comment (trickypr) [inline src/browser/base/content/browser-sidebar-js.patch]: Is there a reason why you aren't using `this.browser` here?

File: src/browser/base/content/browser-sidebar-js.patch [hunk 3]
```diff
@@ -84,16 +84,26 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
...
+@@ -76,8 +114,18 @@ var SidebarUI = {
+     }
+     return (this.__title = document.getElementById("sidebar-title"));
    },
++
++  get keepTabsActivePref() {
++    if (this._keepTabsActive != null) {
++      return this._keepTabsActive;
++    }
++    return this._keepTabsActive = Services.prefs.getBoolPref("pulse.sidebar.keeptabsactive.enabled", true);
```
Comment (trickypr) [inline src/browser/base/content/browser-sidebar-js.patch]: Why is there not an observer to detect when this preference changes and reset everything? Have I missed it?

File: src/browser/locales/en-US/browser/preferences/preferences-ftl.patch [hunk 4]
```diff
@@ -11,6 +11,8 @@ index b10403800c7aafe97471474f6bddb6098ce21a82..b39593816149a71a1e6b40626e304e36
 +preferences-sidebar-header = Sidebar
 +preferences-sidebar-visible =
 +    .label = Show sidebar tabs 
-+
 +preferences-sidebar-extensions-visible =
 +    .label = Show available extensions on the sidebar
++preferences-sidebar-active-tabs = 
++    .label = Keep sidebar tabs active even when closed
++preferences-sidebar-active-tabs-tooltip = When enabled, sidebar tabs will remain active even when closed. When disabled, sidebar tabs will be closed when you click on another tab. <a data-l10n-name="restart-browser">Restart { -brand-short-name }</a>
```
Comment (trickypr) [inline src/browser/locales/en-US/browser/preferences/preferences-ftl.patch]: Preference changes do not require restarts. If they do, you have written incorrect code.

File: src/browser/components/preferences/main-inc-xhtml.patch [hunk 5]
```diff
@@ -13,6 +13,13 @@ index 3d4035690d38f4d851cc86749ebeec7b7ca9aaf2..e9136a52427d7f3ada87803373f16c20
 +  <checkbox id="sidebarVisible"
 +    preference="pulse.sidebar.enabled"
 +    data-l10n-id="preferences-sidebar-visible"/>
++  <checkbox id="sidebarKeepTabsActive"
++    preference="pulse.sidebar.keeptabsactive.enabled"
++    data-l10n-id="preferences-sidebar-active-tabs"/>
++  <description id="sidebarKeepTabsActiveTooltip"
```
Comment (trickypr) [inline src/browser/components/preferences/main-inc-xhtml.patch]: This is a description (mozilla calls it a description or footer depending on the circumstance), not a tooltip.

**Review by trickypr — CHANGES_REQUESTED** [review]
This will not be merged if it requires a browser restart to work correctly. Either:
a) Remove the preference
b) Add a preference observer and fix whatever ends up being broken that requires a browser restart

File: src/browser/base/content/browser-sidebar-js.patch [hunk 6]
```diff
@@ -165,7 +205,89 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
...
-@@ -573,6 +652,10 @@ var SidebarUI = {
+@@ -485,6 +588,11 @@ var SidebarUI = {
+   },
+ 
+   _loadSidebarExtension(commandID) {
++    if (this.keepTabsActivePref && this.loadedSidebars.includes(commandID)) {
++      console.log("returned");
++      return;
++    }
++    console.log("it went on");
```
Comment (trickypr) [inline src/browser/base/content/browser-sidebar-js.patch]: Also, you still have debug console.log statements here

File: src/browser/base/content/browser-sidebar-js.patch [hunk 7]
```diff
@@ -134,7 +144,37 @@ index f09e37fc42f7754f7f3687884ead25124c1b2076..e1833bbfd8a8ed7da767349f886d554f
      this._inited = true;
  
      Services.obs.addObserver(this, "intl:app-locales-changed");
-@@ -159,17 +229,26 @@ var SidebarUI = {
+ 
+     this._initDeferred.resolve();
++    
++    //check if element by id sidebar exists to bool var
```
Comment (trickypr) [inline src/browser/base/content/browser-sidebar-js.patch]: What is this comment? It doesn't appear to add much of value

Comment (PressJump) [reply src/browser/base/content/browser-sidebar-js.patch]: Probably remnant of testing and etc that I forgot to wipe, not ready to review yet.

File: src/browser/base/content/browser-sidebar-js.patch (hunk 6 above)
Comment (PressJump) [reply src/browser/base/content/browser-sidebar-js.patch]: Also left overs from testing which I forgot to wipe.

**Comment by PressJump** [comment]
Changed to draft until changes made, I am not too sure why it was not on draft already after last review.

**Comment by PressJump** [comment]
Bug where a sidebar item gets launched into the built in icons after disabling the keep sidebar items active. 
image

I believe this is caused because we are rerunning initialization.

**Comment by PressJump** [comment]
Preference now relies on observer, reloading the sidebar and removing any left over content and changes have been updated to sync with alpha.

File: src/browser/components/preferences/main-js.patch [hunk 8]
```diff
@@ -1,15 +1,17 @@
...
-@@ -161,6 +161,11 @@ Preferences.addAll([
+@@ -161,6 +161,13 @@ Preferences.addAll([
  
    // Media
    { id: "media.hardwaremediakeys.enabled", type: "bool" },
 +
 +
 +  // Pulse browser: custom settings
 +  { id: 'pulse.sidebar.enabled', type: 'bool' },
++  { id: 'pulse.sidebar.extensions.enabled', type: 'bool' },
```
Comment (trickypr) [inline src/browser/components/preferences/main-js.patch]: Why is the extension preference back?

**Review by trickypr — CHANGES_REQUESTED** [review]

**Review by PressJump — COMMENTED** [review]
Yea sorry, forgot to lint.

**Review by trickypr — APPROVED** [review]

**Pull request merged**